### ISSUE TITLE
set SCRIPT_NAME (properly?) in CGI environment

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2075]
|<img alt="martinpaljak's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| martinpaljak reported|
|-|-|
|Trac ID|trac#2075|
|Type|defect|
|Created|2006-09-10 22:02:18Z|

http://cgi-spec.golux.com/draft-coar-cgi-v11-03-clean.html tells:

  The SCRIPT_NAME metavariable is set to a URL path that could identify the CGI script (rather than the script's output). The syntax and semantics are identical to a decoded HTTP URL 'path' token (see RFC 2396 [4]).
```
    SCRIPT_NAME = "" | ( "/" [ path ] )
```

PEP-333 tells:
  The following variables must be present, unless their value would be an empty string, in which case they may be omitted, except as otherwise noted below.

and 

  SCRIPT_NAME
  The initial portion of the request URL's "path" that corresponds to the application object, so that the application knows its virtual "location". This may be an empty string, if the application corresponds to the "root" of the server.

When putting together URLMap from paste (http://pythonpaste.org/module-paste.urlmap.html) in a root WSGIResource all URLs started to contain double slash in the beginnig of relativ URLs generated by mapped apps and this caused URLMap to fail.

This patch fixes the symptoms and i think is anyway the right thing to do.

```
Index: twisted/web2/twcgi.py
===================================================================
--- twisted/web2/twcgi.py       (revision 18085)
+++ twisted/web2/twcgi.py       (working copy)
@@ -59,8 +59,11 @@
     env["REMOTE_ADDR"] = remotehost.host
     env["REQUEST_METHOD"] = request.method
     # Should we raise an exception if this contains "/" chars?
-    env["SCRIPT_NAME"] = '/' + '/'.join(request.prepath)
-    
+    if request.prepath:
+           env["SCRIPT_NAME"] = '/' + '/'.join(request.prepath)
+    else:
+       env["SCRIPT_NAME"] = ''
+
     env["SERVER_NAME"] = request.host
     env["SERVER_PORT"] = str(request.port)
     
```

Attachments:

* [wsgi_test_SCRIPT_NAME.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/884/884740bce764456d0d18e701f5d60563c452e663/b78b1bb00da6456b5d89ea3491f8d6b4695e0edf.diff) (528 bytes) - added by karlb on 2007-04-30 13:57:36Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2075 2075
type__defect defect
reporter__martinpaljak martinpaljak
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1157925738000000 1157925738000000
changetime__1181940715000000 1181940715000000
version__None None
owner__dialtone dialtone
cc__jknight cc__martinpaljak cc__karlb cc__exarkun cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml set owner to martinpaljak|
|-|-|

This definitely needs a test before it can be included. At the least, the ticket should show a clear way to reproduce the bug -- a unit test is a great way to do this.

**Comment by twisted-trac** [comment]
|<img alt="martinpaljak's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|martinpaljak commented|
|-|-|

I would like to ask for opinions about interpreting different standards and selecting the 'right' implementation. IMHO the change makes it 'right' but there might be other opinions. If the change is considered to be the 'right' way, then i can write a test case to demonstrate it, currently i can only assert that this way is more right than the current implementation and create a test case that demonstrates a different outcome than what i want (what does not render current functionality a bug). As you can see, all important words are 'may' not 'should' or 'must'.

The important part again:
```
This may be an empty string, if the application corresponds to the "root" of the server.
```

This is exactly the case for me (the root object is a wsgi resource) so the variable MAY be an empty string what exactly this patch makes it to be. And it makes other software work properly. I'm not in the position to decide if this MAY is the right way or not.

For evaluation, please see http://twistedmatrix.com/trac/browser/trunk/twisted/web2/twcgi.py#L53 - This would be in sync with this patch and combines in the 'may be empty' clause

**Comment by twisted-trac** [comment]
|<img alt="martinpaljak's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|martinpaljak set owner to @jyknight|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_6"></a>|@jyknight commented|
|-|-|

What you propose makes a lot of sense, but what would really convince me is to check to see what other HTTP servers (say, apache) do with CGI scripts at the root.

**Comment by twisted-trac** [comment]
|<img alt="martinpaljak's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|martinpaljak commented|
|-|-|

Replying to [jknight](#note_7):

I think that it doesn't make sense in case of normal servers and CGI scripts - with CGI scripts you usually always have a real script. I really don't know how to create a similar environment in apache using cgi or php for example - there would always be a real file somewhere that it would be mapped to. So as the named functionality in twisted is borrowed from CGI and uses CGI specs that indeed, have SOME script somewhere that this variable can be mapped to, but in the context of WSGI it may not be so (as in my case). Also, the title is a bit confusing - i mixed cgi with wsgi, when the problem only appears in wsgi.

As a reference, wsgiref uses empty string (and that should be the reference wsgi implementation :))

So there are two options:
 * change it in twcgi.py and assume that in case of normal cgi scripts this condition (prepath being empty) is never met
 * overwrite wsgi environment creation for  SCRIPT_NAME

I suggest the first option

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_10"></a>|@jyknight set owner to martinpaljak|
|-|-|

I agree with the first option (as written originally), but not because it'll never be met, but because you've convinced me it's probably correct even if it is. :) Please write a test case and I'll merge it.

**Comment by twisted-trac** [comment]
|<img alt="martinpaljak's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|martinpaljak commented|
|-|-|

Replying to [jknight](#note_10):
> I agree with the first option (as written originally), but not because it'll never be met, but because you've convinced me it's probably correct even if it is. :) Please write a test case and I'll merge it.

I gave it a quick look but could not figure out how to trick test_wsgi and the assertEnv method to deal properly with the prepath - i could not generate a useful case at least.. :(

**Comment by twisted-trac** [comment]
|<img alt="karlb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|karlb commented|
|-|-|

While trying to run pylons on top of web2, I also stumbled upon this bug. Routes (one of the parts of a standard pylons setup) is irritated by the '/' SCRIPT_PATH and starts to generate wrong URLs. The patch fixes all problems and works fine.

**Comment by twisted-trac** [comment]
|<img alt="karlb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|karlb commented|
|-|-|

I attached the missing test. The test only works after appying [#2612](https://github.com/twisted/twisted/issues/2612).

**Comment by twisted-trac** [comment]
|<img alt="martinpaljak's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|martinpaljak commented|
|-|-|

Nice. We have the test as well now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to karlb|
|-|-|

I applied the test patch and ran it, but it doesn't fail, even without the change to createCGIEnvironemnt.  Is the test wrong, or did someone else accidentally fix this?

**Comment by twisted-trac** [comment]
|<img alt="karlb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|karlb commented|
|-|-|

As noted above the test only works after applying [[#2612](https://github.com/twisted/twisted/issues/2612)](https://github.com/twisted/twisted/issues/2612) (otherwise it always succeeds). I can create a patch that works alone, but it would either duplicate a few line of code or conflict with #2612.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

Ah, I didn't succeed in interpreting "only works" as "correctly fails without the presence of the fix".

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve set owner to @therve|
|-|-|

I guess this one should wait for [#2612](https://github.com/twisted/twisted/issues/2612).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve set owner to @jyknight|
|-|-|

Ready to review in cgi-script-name-2075.

**Comment by twisted-trac** [comment]
|<img alt="dialtone's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_24"></a>|dialtone commented|
|-|-|

Did a review and added some more tests in the branch.

Ready for a second review from someone else.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_25"></a>|@jyknight set owner to dialtone|
|-|-|

Please add a docstring to assertEnv, since it's being modified; other than that, looks good to me.

**Comment by twisted-trac** [comment]
|<img alt="dialtone's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_26"></a>|dialtone set status to `closed`|
|-|-|

(In [20548]) Merge cgi-script-name-2075: SCRIPT_NAME env variable correctly added to CGI scripts

Added a test and fixed the problem where twisted.web2 always used some
value for the SCRIPT_NAME env var even when it shouldn't have.

Author: dialtone
Reviewer: foom
Fixed [#2075](https://github.com/twisted/twisted/issues/2075)
